In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

from keras.layers.core import Dense, Activation, Dropout
from keras.layers.recurrent import LSTM
from keras.models import Sequential
%matplotlib inline

# Display only top 5 and bottom 5 rows
pd.set_option('display.max_rows', 10)

Using TensorFlow backend.


In [2]:
# Set random seed for reproducibility
# Note: CuDNN is usually non-deterministic
# (can't determine which of the ~3000 threads finish earlier)
# And floating points reduction is not perfectly associative due to ULP rounding
import numpy as np
np.random.seed(1337)
import tensorflow as tf
tf.set_random_seed(1337)

In [10]:
df_train = pd.read_csv('./data/train_aWnotuB.csv', parse_dates=[0], infer_datetime_format=True)
df_train

DateTime  Junction  Vehicles           ID
0     2015-11-01 00:00:00         1        15  20151101001
1     2015-11-01 01:00:00         1        13  20151101011
2     2015-11-01 02:00:00         1        10  20151101021
3     2015-11-01 03:00:00         1         7  20151101031
4     2015-11-01 04:00:00         1         9  20151101041
...                   ...       ...       ...          ...
48115 2017-06-30 19:00:00         4        11  20170630194
48116 2017-06-30 20:00:00         4        30  20170630204
48117 2017-06-30 21:00:00         4        16  20170630214
48118 2017-06-30 22:00:00         4        22  20170630224
48119 2017-06-30 23:00:00         4        12  20170630234

[48120 rows x 4 columns]

In [11]:
train = df_train.pivot(index='DateTime',columns='Junction', values='Vehicles')
train

Junction                 1     2     3     4
DateTime                                    
2015-11-01 00:00:00   15.0   6.0   9.0   NaN
2015-11-01 01:00:00   13.0   6.0   7.0   NaN
2015-11-01 02:00:00   10.0   5.0   5.0   NaN
2015-11-01 03:00:00    7.0   6.0   1.0   NaN
2015-11-01 04:00:00    9.0   7.0   2.0   NaN
...                    ...   ...   ...   ...
2017-06-30 19:00:00  105.0  34.0  33.0  11.0
2017-06-30 20:00:00   96.0  35.0  31.0  30.0
2017-06-30 21:00:00   90.0  31.0  28.0  16.0
2017-06-30 22:00:00   84.0  29.0  26.0  22.0
2017-06-30 23:00:00   78.0  27.0  39.0  12.0

[14592 rows x 4 columns]

### Remove Nan (0 vehicle)

In [17]:
train = train.fillna(0)

# Generate rolling forecast features

In [18]:
nb_forecast_per_junction = 24 * (31 + 31 + 30 + 31) # Days in jul + aug + sep + oct

In [19]:
nb_forecast_per_junction

2952

In [20]:
nb_forecast_per_junction * 4

11808

That will certainly not fit in the GPU VRAM --> We will get inspiration from seq2seq models and do a sliding window of time that matches a week.

In [8]:
num_features = 4
seq_len = 24 * 7 # Sequence of one week of traffic: 168
num_outputs = 4
num_hidden = 52 * 1.5 # Keep 1.5 years as hidden state
bs = 16

We originally had 14592 rows, we will generate 168 sequence of prediction from it.

In [42]:
def make_input_seqs(data, seq_len, train_split=0.9):
    seq_len = seq_len + 1
    result = []
    for index in range(len(data) - seq_len):
        result.append(data[index: index + seq_len, :])
    result = np.array(result) # shape (14423, 168, 4)
    train_ind = round(train_split * result.shape[0])
    train = result[:int(train_ind), :, :]
    x_train = train[:, :-1, :]
    y_train = train[:, -1, :]
    x_test = result[int(train_ind):, :-1, :]
    y_test = result[int(train_ind):, -1, :]

    return [x_train, y_train, x_test, y_test]

In [43]:
X_train, y_train, X_test, y_test = make_input_seqs(train.values, seq_len)